infrastructure/src/main/java/org/corfudb/infrastructure/management/ClusterStateContext.java

```diff
@@ -1,199 +1,62 @@
-     * @return True if the node state is stale. False otherwise.
-     */
-    private boolean isNodeStateStale(NodeState nodeState) {
-        final NodeHeartbeat nodeHeartbeat
-                = clusterView.get(NodeLocator.getLegacyEndpoint(nodeState.getEndpoint()));
-        return nodeHeartbeat != null && getHeartbeatFromNodeState(nodeState)
-                .compareTo(getHeartbeatFromNodeState(nodeHeartbeat.getNodeState())) <= 0;
-    }
-
-    /**
-     * Updates the a particular node state in the cluster view.
-     *
-     * @param endpoint  Endpoint of whose node state is to be updated.
-     * @param nodeState Node state to be updated.
-     */
-    private void updateNodeState(@NonNull String endpoint, @NonNull NodeState nodeState, @NonNull Layout layout) {
-        clusterView.compute(endpoint, (s, nodeHeartbeat) ->
-                nodeHeartbeat != null && isNodeStateStale(nodeState) ? nodeHeartbeat
-                        : new NodeHeartbeat(nodeState,
-                        new HeartbeatTimestamp(layout.getEpoch(), heartbeatCounter)));
-    }
-
-    /**
-     * Updates the cluster view based on another cluster view received in the heartbeat response.
-     * The latest Node states from this cluster state are retained and the stale ones are
-     * discarded.
-     *
-     * @param clusterState Cluster state to update the local copy of the cluster state with.
-     */
-    public synchronized void updateNodeState(@NonNull ClusterState clusterState, @NonNull Layout currentLayout) {
-        clusterState.getNodeStatusMap().forEach((endpoint, nodeState) ->
-                updateNodeState(endpoint, nodeState, currentLayout));
-    }
-
-    /**
-     * Update the decay flags in the NodeHeartbeats if the cluster state contains a stale NodeState.
-     */
-    private void markDecayedNodeStates(Layout layout, final long heartbeatCounter) {
-
-        clusterView.forEach((endpoint, nodeHeartbeat) -> {
-            HeartbeatTimestamp currentTimestamp = new HeartbeatTimestamp(layout.getEpoch(), heartbeatCounter);
-            HeartbeatTimestamp lastSeen = nodeHeartbeat.getLocalHeartbeatTimestamp();
-            if (currentTimestamp.epoch > lastSeen.epoch
-                    || (currentTimestamp.epoch == lastSeen.epoch
-                    && currentTimestamp.counter > lastSeen.counter + NODE_STATE_DECAY_CONSTANT)) {
-                nodeHeartbeat.setDecayed(true);
-            }
-        });
-    }
+    @NonNull
+    private final AtomicReference<ClusterState> clusterView;
+    @Getter
+    @NonNull
+    private final HeartbeatCounter counter;
 
     /**
      * Refreshes the cluster view based on the local endpoint and the snapshot epoch at which the
      * node states are updated.
      *
-     * @param localEndpoint  Local Endpoint whose node state entry is to be updated in the cluster
-     *                       view map.
-     * @param snapshotLayout Snapshot layout.
+     * @param layout Snapshot layout.
```

Comment:
Miss the description of report in the comment.

Reply:
done